### PR TITLE
get nix-writers into nixpkgs

### PR DESCRIPTION
**Pull request by Lassulus**
Get https://cgit.krebsco.de/nix-writers/about/ into nixpkgs

feedback is appreciated

stuff which still needs to be done:

- [x] comment all the functions in pkgs/build-support/writers.nix
- [x] unify interface between writeJS, writeC, writeHaskellPackage and writePerl (deps vs dependencies vs libraries)

nice to have:

- [x] writeHaskellPackage dependencies are attributes instead of strings
- [x] writeC dependencies are a list instead of attrset

### CODE REVIEW

**Comment by matthewbauer** [comment]
I think something like this needs some good use cases to be included in full. Right now a lot of them seem like they have a pretty limited use case. I'd definitely be more comfortable with some getting included if it leads to less total code in Nixpkgs.

I had a similar try at something like ```writers.write``` in #43074. ```buildTree``` does I think the same thing so that these two are equivalent:

```
write "name" { "/etc/test" = { text = "hello world"; }; }
```

```
buildTree "name" {
  "/etc/test" = writeText "test" "hello world";
} {}
```

So I think something like that is definitely useful.

**Comment by Lassulus** [comment]
I think most of the use-cases would be outside of nixpkgs or in nixos modules.
It's very cool to have inline packages in your shell.nix or ExecStart scripts in other languages than bash.

File: pkgs/build-support/writers.nix [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  # Base implementation for non-compiled executables.
+  # Takes an interpreter, for example `${pkgs.bash}/bin/bash`
+  #
+  # Examples:
+  #   writebash = makeScriptWriter { interpreter = "${pkgs.bash}/bin/bash"; }
+  #   makeScriptWriter { interpreter = "${pkgs.dash}/bin/dash"; } "hello" "echo hello world"
+  makeScriptWriter = { interpreter, check ? null }: name: text:
+    assert (with types; either absolute-pathname filename).check name;
+    pkgs.write (baseNameOf name) {
+      ${optionalString (types.absolute-pathname.check name) name} = {
```
Comment (Mic92) [inline pkgs/build-support/writers.nix]: Is the syntax here correct?

File: lib/types.nix [hunk 2]
```diff
@@ -243,6 +243,43 @@ rec {
       merge = mergeOneOption;
     };
 
+    posix = {
```
Comment (edolstra) [inline lib/types.nix]: IMHO these `posix` types should be a separate PR. At first glance, they strike me as a bit overengineered and not really necessary to the writers stuff.

Comment (Lassulus) [reply lib/types.nix]: Alright, I removed them

**Comment by Mic92** [comment]
We have `pkgs/test`, but maybe something in the writers directory would make more sense.

**Comment by Lassulus** [comment]
Soo I guess I'm done. I'm not sure about the tests though

**Comment by matthewbauer** [comment]
I think we need better criteria of where this can be used. Is this intended for internal use within Nixpkgs? If so, we need some usages inside the tree. If not, why not just release this as a separate repo? I just don't think we have a good enough use for this right now.

Also, It also doesn't really support external files, which seems pretty important. You can kind of "fake" it with something like:

```
writeBash "examples" (builtins.readFile ./example.sh)
```

But this just reads the file and writes it back to the store twice! Anyway, this seems like something that is an interesting use of Nix, but should be left out of Nixpkgs.

**Comment by Lassulus** [comment]
There are usecases inside of nixpkgs. Having something like services.xserver.windowManager.xmonad.config comes to my mind. I can implement this next. Also a lot of suckless software rely upon recompliation and there is no abstraction inside configuration.nix

About the files ending up twice in the store, does writeScript(Bin) handle this?
I guess, I could extend the writers to take strings or files.
